source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,927 @@
+error if there is a timeout on the session but also an overridden timeout for
+the operation. It’s possible that the ability to communicate timeouts for a
+block of operations via a ClientSession is useful as a general purpose API,
+but we’ve decided to make it private until there are other known use cases.
+
+withTransaction refreshes the timeout for abortTransaction
+----------------------------------------------------------
+
+If the user-provided callback to ``withTransaction`` times out, it could
+leave a transaction running on the server. It’s imperative that drivers make
+an effort to abort the open transaction because failing to do so could result
+in the collections and databases affected by the transaction being locked for
+a long period of time, which could cause applications to stall. Because
+``timeoutMS`` has expired before drivers attempt to abort the transaction, we
+require drivers to refresh it and apply the original value to the execution
+of the ``abortTransaction`` operation. This can cause the entire
+``withTransaction`` call to take up to ``2*timeoutMS``, but it was decided
+that this risk is worthwhile given the importance of transaction cleanup.
+
+GridFS streams behavior
+-----------------------
+
+Streams created by GridFS API operations (e.g. by ``open_upload_stream`` and
+``open_download_stream``) present a challenge for this specification. These
+types of streams execute multiple operations, but there can be artificial
+gaps between operations if the application does not invoke the stream
+functions for long periods of time. Generally, we expect users to upload or
+download an entire file as quickly as possible, so we decided to have
+``timeoutMS`` cap the lifetime of the created stream. The other option was to
+apply the entire ``timeoutMS`` value to each operation executed by the
+stream, but streams perform many hidden operations, so this approach could
+cause an upload/download to take much longer than expected.
+
+timeoutMS cannot be overridden for StartSession calls
+-----------------------------------------------------
+
+In general, users can override ``timeoutMS`` at the level of a single
+operation.  The ``StartSession`` operation, however, only inherits
+``timeoutMS`` from the MongoClient and does not allow the option to be
+overridden.  This was a consious API design decision because drivers are
+moving towards only supporting MongoDB versions 3.6 and higher, so sessions
+will always be supported. Adding an override for ``StartSession`` would
+introduce a new knob and increase the API surface of drivers without providing
+a significant benefit.
+
+
+Future work
+===========
+
+Modify GridFS streams behavior via new options
+----------------------------------------------
+
+As explained in the design rationale, drivers use ``timeoutMS`` to cap the
+entire lifetime of streams created by GridFS operations such as
+``open_upload_stream`` and ``open_download_stream``. If we find that users are
```

Comment:
Per our conversation yesterday, the "LIFETIME" scope of `timeoutMS` for GridFS wasn't just about the methods that return stream objects for applications to interact with directly. It equally applies to the streams created internally for operations such as `downloadToStream`.

I'd suggest removing "such as..." from this sentence or clarifying that it applies equally to operations that return a stream and those that do no (since they perform all of the GridFS IO internally).

Reply:
done.